### PR TITLE
Initial Firestore plugin implementation

### PR DESCRIPTION
**Pull request by collinjackson**
Fixes flutter/flutter#12371

### CODE REVIEW

File: packages/firestore/LICENSE [hunk 1]
```diff
@@ -0,0 +1,26 @@
+Copyright 2017, the Flutter project authors. All rights reserved.
```
Comment (goderbauer) [inline packages/firestore/LICENSE]: Unify copyright lines? Here and in the source file headers?

Some source file headers say "Chromium authors", which is what we use for the other plugins as well, I think.

File: packages/firestore/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    defaultConfig {
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+    lintOptions {
+        disable 'InvalidPackage'
+    }
+    dependencies {
+        compile 'com.google.firebase:firebase-firestore:11.4.2'
+//        compile 'com.google.firebase:firebase-core:11.0.+'
```
Comment (goderbauer) [inline packages/firestore/android/build.gradle]: Any reason why these are commented out? Maybe just remove them?

File: packages/firestore/lib/src/firestore.dart [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_firestore;
+
+/// The entry point for accessing a Firebase Database. You can get an instance
+/// by calling `FirebaseDatabase.instance`. To access a location in the database
```
Comment (goderbauer) [inline packages/firestore/lib/src/firestore.dart]: Should this be Firestore.instance?

File: packages/firestore/lib/src/firestore.dart [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_firestore;
+
+/// The entry point for accessing a Firebase Database. You can get an instance
```
Comment (goderbauer) [inline packages/firestore/lib/src/firestore.dart]: Is Firebase Database correct? Or should this be the Firestore?

File: packages/firestore/lib/src/firestore.dart [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+/// by calling `FirebaseDatabase.instance`. To access a location in the database
+/// and read or write data, use `reference()`.
```
Comment (goderbauer) [inline packages/firestore/lib/src/firestore.dart]: This class doesn't have a reference method?

File: packages/firestore/lib/src/firestore.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  static final Map<int, StreamController<QuerySnapshot>> _queryObservers =
+      <int, StreamController<QuerySnapshot>>{};
+
+  static final Map<int, StreamController<DocumentSnapshot>> _documentObservers =
+      <int, StreamController<DocumentSnapshot>>{};
+
+  Firestore._() {
+    channel.setMethodCallHandler((MethodCall call) {
+      if (call.method == 'QuerySnapshot') {
+        print('HEY!! ${call.method} ${call.arguments}');
```
Comment (goderbauer) [inline packages/firestore/lib/src/firestore.dart]: remove prints?

File: packages/firestore/lib/src/document_reference.dart [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_firestore;
+
+/// A DocumentReference refers to a document location in a Firestore database
+/// and can be used to write, read, or listen to the location.
+///
+/// The document at the referenced location may or may not exist.
+/// A DocumentReference can also be used to create a CollectionReference
```
Comment (goderbauer) [inline packages/firestore/lib/src/document_reference.dart]: Use `[]` for CollectionReference to get a free link in the docs?

File: packages/firestore/lib/firebase_firestore.dart [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
```
Comment (goderbauer) [inline packages/firestore/lib/firebase_firestore.dart]: Previous source files where copyrighted to the Chromium Authors. Unify?

File: packages/firestore/lib/src/collection_reference.dart [hunk 9]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_firestore;
+
+/// A CollectionReference object can be used for adding documents, getting
+/// document references, and querying for documents (using the methods
+/// inherited from Query).
```
Comment (goderbauer) [inline packages/firestore/lib/src/collection_reference.dart]: `[Query]` to link to that class?

File: packages/firestore/lib/src/collection_reference.dart [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  /// For root collections, null is returned.
+  CollectionReference parent() {
+    if (_pathComponents.isEmpty) {
+      return null;
+    }
+    return new CollectionReference._(
+        _firestore, (new List<String>.from(_pathComponents)..removeLast()));
+  }
+
+  String get key => _pathComponents.last;
```
Comment (goderbauer) [inline packages/firestore/lib/src/collection_reference.dart]: Document this public member?

File: packages/firestore/lib/src/query.dart [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_firestore;
+
+/// Represents a query over the data at a particular location.
+class Query {
+  Query._({
+    @required Firestore firestore,
+    @required List<String> pathComponents,
```
Comment (goderbauer) [inline packages/firestore/lib/src/query.dart]: Is it ok for the required pathComponents to be empty or null? Maybe add an assert just like it was added for firestore?

File: packages/firestore/lib/src/query.dart [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    return new Query._(
+      firestore: _firestore,
+      pathComponents: _pathComponents,
+      parameters: new Map<String, dynamic>.unmodifiable(
+        new Map<String, dynamic>.from(_parameters)..addAll(parameters),
+      ),
+    );
+  }
+
+  Map<String, dynamic> buildArguments() {
```
Comment (goderbauer) [inline packages/firestore/lib/src/query.dart]: Document public memebers?

File: packages/firestore/lib/src/utils/push_id_generator.dart [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+///
+/// Since the Flutter plugin API is asynchronous, there's no way for us
+/// to use the native SDK to generate the node key synchronously and we
+/// have to do it ourselves if we want to be able to reference the
+/// newly-created node synchronously.
+///
+/// This code is based on a Firebase blog post and ported to Dart.
+/// https://firebase.googleblog.com/2015/02/the-2120-ways-to-ensure-unique_68.html
+class PushIdGenerator {
+  static const String PUSH_CHARS =
```
Comment (goderbauer) [inline packages/firestore/lib/src/utils/push_id_generator.dart]: Can this be private?

File: packages/firestore/lib/src/utils/push_id_generator.dart [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+      '-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz';
+
+  static final Random _random = new Random();
+
+  static int _lastPushTime;
+
+  static final List<int> _lastRandChars = new List<int>(12);
+
+  static String generatePushChildName() {
```
Comment (goderbauer) [inline packages/firestore/lib/src/utils/push_id_generator.dart]: Document public members?

Comment (collinjackson) [reply packages/firestore/lib/src/utils/push_id_generator.dart]: This is not a public API, it is only used within the package.

File: packages/firestore/lib/src/utils/push_id_generator.dart (hunk 13 above)
Comment (collinjackson) [reply packages/firestore/lib/src/utils/push_id_generator.dart]: This is not a public API, it is only used within the package.

Comment (collinjackson) [reply packages/firestore/lib/src/utils/push_id_generator.dart]: I could make it an underscore class and use `part` if you think that's better...

Comment (goderbauer) [reply packages/firestore/lib/src/utils/push_id_generator.dart]: Oh, you're right. For some reason I though this was brought into the global name space with part of. But you're right, it is not.

File: packages/firestore/lib/src/query.dart (hunk 11 above)
Comment (collinjackson) [reply packages/firestore/lib/src/query.dart]: Done

File: packages/firestore/lib/src/collection_reference.dart (hunk 10 above)
Comment (collinjackson) [reply packages/firestore/lib/src/collection_reference.dart]: Good catch, it should actually be path. I'll fix.

File: packages/firestore/lib/src/collection_reference.dart (hunk 9 above)
Comment (collinjackson) [reply packages/firestore/lib/src/collection_reference.dart]: Done

File: packages/firestore/lib/firebase_firestore.dart (hunk 8 above)
Comment (collinjackson) [reply packages/firestore/lib/firebase_firestore.dart]: Done

File: packages/firestore/lib/src/document_reference.dart (hunk 7 above)
Comment (collinjackson) [reply packages/firestore/lib/src/document_reference.dart]: Done

File: packages/firestore/lib/src/firestore.dart (hunk 5 above)
Comment (collinjackson) [reply packages/firestore/lib/src/firestore.dart]: Fixed

File: packages/firestore/lib/src/firestore.dart (hunk 3 above)
Comment (collinjackson) [reply packages/firestore/lib/src/firestore.dart]: Fixed

File: packages/firestore/lib/src/firestore.dart (hunk 4 above)
Comment (collinjackson) [reply packages/firestore/lib/src/firestore.dart]: It's a mistake, thanks.

File: packages/firestore/LICENSE (hunk 1 above)
Comment (collinjackson) [reply packages/firestore/LICENSE]: Fixed

File: packages/firestore/README.md [hunk 15]
```diff
@@ -10,7 +10,7 @@ For Flutter plugins for other Firebase products, see [FlutterFire.md](https://gi
 
 To use this plugin:
 
-1. Using the [Firebase Console](http://console.firebase.google.com/), add an Android app to your project: Follow the assistant, download the generated google-services.json file and place it inside android/app. Next, modify the android/build.gradle file and the android/app/build.gradle file to add the Google services plugin as described by the Firebase assistant.
+1. Using the [Firebase Console](http://console.firebase.google.com/), add an Android app to your project: Follow the assistant, download the generated google-services.json file and place it inside android/app. Next, modify the android/build.gradle file and the android/app/build.gradle file to add the Google services plugin as described by the Firebase assistant. Ensure that your `android/build.gradle` file contains the `maven.google.com` as [described here](https://firebase.google.com/docs/android/setup#add_the_sdk).
```
Comment (goderbauer) [inline packages/firestore/README.md]: non-actionable comment: any flutter app that was created within the last 4 month already includes this one per default (after creating it with `flutter create`).

**Pull request merged**